### ISSUE TITLE
Perl_magic_setsig/clearsig problems (patch included)

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6260]
Migrated from [rt.perl.org#20613](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=20613) (status was 'resolved')

Searchable as RT20613$

**Comment by p5pRT** [comment]
### From ajohnson@wischip.com

I discovered a few problems with setting and clearing %SIG entries in
the Perl 5\.8\.0 production release \(currently the same as the development
release\)\. To wit&#8203;:
 
\+ Deleting a %SIG entry does not restore the default signal/hook
disposition\.
\+ The previous disposition SV refcount is decremented \(possibly
resulting in its deallocation\) \*before\* the new disposition is in place\,
and while the handler is in an inconsistent state\. If it's a coderef
with closures\, then just about anything can happen at that point\, which
is bad because DESTROY might need the new handler to be in place\.
\+ A signal can arrive while it's being dispositioned\, which could\, for
example\, cause the program to terminate immediately while we're
switching from one handler to another\.
\+ There are a couple of apparently redundant "\*svp=0" statements\. \(I
commented them\, but I was too chicken to remove them\.\)
 
The patch includes a test case for the first 2 items\. Here are a couple
of items that I might have screwed up and should be reviewed by a more
experienced Perl hacker&#8203;:
 
\+ The test case uses "$? & 0xFF" to see if the closed pipe exited from a
signal\. Is this portable?
\+ Is there any way \(e\.g\. exception?\) that Perl\_magic\_setsig\(\) might
escape without restoring the sigprocmask? \(If so\, we could maybe use the
trick from POSIX&#8203;::sigaction to get around that\.\) This could also result
in a memory leak \(because "to\_dec" would get ignored\)\.
 
Here is the patch&#8203;:
 
<details><summary>Inline Patch</summary>

```diff
diff -Naur perl-5.8.0/mg.c perl-5.8.0-sig2/mg.c
--- perl-5.8.0/mg.c          Sat Jun 15 13:16:44 2002
+++ perl-5.8.0-sig2/mg.c            Wed Jan 29 13:27:34 2003
@@ -1061,19 +1061,48 @@
 int
 Perl_magic_clearsig(pTHX_ SV *sv, MAGIC *mg)
 {
-    I32 i;
+    /* XXX Some of this code was copied from Perl_magic_setsig. A

```
</details>
little
\+     \* refactoring might be in order\.
\+     \*/
\+    register char \*s;
&nbsp;     STRLEN n\_a;
\-    /\* Are we clearing a signal entry? \*/
\-    i = whichsig\(MgPV\(mg\,n\_a\)\);
\-    if \(i\) \{
\-           if\(PL\_psig\_ptr\[i\]\) \{
\-               SvREFCNT\_dec\(PL\_psig\_ptr\[i\]\);
\-               PL\_psig\_ptr\[i\]=0;
\-           \}
\-           if\(PL\_psig\_name\[i\]\) \{
\-               SvREFCNT\_dec\(PL\_psig\_name\[i\]\);
\-               PL\_psig\_name\[i\]=0;
\-           \}
\+    SV\* to\_dec;
\+    s = MgPV\(mg\,n\_a\);
\+    if \(\*s == '\_'\) \{
\+          SV\*\* svp;
\+          if \(strEQ\(s\,"\_\_DIE\_\_"\)\)
\+              svp = &PL\_diehook;
\+          else if \(strEQ\(s\,"\_\_WARN\_\_"\)\)
\+              svp = &PL\_warnhook;
\+          else
\+              Perl\_croak\(aTHX\_ "No such hook&#8203;: %s"\, s\);
\+          if \(\*svp\) \{
\+              to\_dec = \*svp;
\+              \*svp = 0;
\+              SvREFCNT\_dec\(to\_dec\);
\+          \}
\+    \}
\+    else \{
\+          I32 i;
\+          /\* Are we clearing a signal entry? \*/
\+          i = whichsig\(s\);
\+          if \(i\) \{
\+\#ifdef FAKE\_DEFAULT\_SIGNAL\_HANDLERS
\+              sig\_defaulting\[i\] = 1;
\+              \(void\)rsignal\(i\, &Perl\_csighandler\);
\+\#else
\+              \(void\)rsignal\(i\, SIG\_DFL\);
\+\#endif
\+              if\(PL\_psig\_name\[i\]\) \{
\+                      SvREFCNT\_dec\(PL\_psig\_name\[i\]\);
\+                      PL\_psig\_name\[i\]=0;
\+              \}
\+              if\(PL\_psig\_ptr\[i\]\) \{
\+                      to\_dec=PL\_psig\_ptr\[i\];
\+                      PL\_psig\_ptr\[i\]=0;
\+                      SvREFCNT\_dec\(to\_dec\);
\+              \}
\+          \}
&nbsp;     \}
&nbsp;     return 0;
 \}
@&#8203;@&#8203; \-1156\,7 \+1185\,15 @&#8203;@&#8203;
&nbsp;     register char \*s;
&nbsp;     I32 i;
&nbsp;     SV\*\* svp = 0;
\+    /\* Need to be careful with SvREFCNT\_dec\(\)\, because that can have
side
\+     \* effects \(due to closures\)\. We must make sure that the new
disposition
\+     \* is in place before it is called\.
\+     \*/
\+    SV\* to\_dec = 0;
&nbsp;     STRLEN len;
\+\#ifdef HAS\_SIGPROCMASK
\+    sigset\_t set\, save;
\+\#endif
 
&nbsp;     s = MgPV\(mg\,len\);
&nbsp;     if \(\*s == '\_'\) \{
@&#8203;@&#8203; \-1168\,7 \+1205\,7 @&#8203;@&#8203;
&nbsp;                Perl\_croak\(aTHX\_ "No such hook&#8203;: %s"\, s\);
&nbsp;            i = 0;
&nbsp;            if \(\*svp\) \{
\-               SvREFCNT\_dec\(\*svp\);
\+              to\_dec = \*svp;
&nbsp;                \*svp = 0;
&nbsp;            \}
&nbsp;     \}
@&#8203;@&#8203; \-1179\,6 \+1216\,12 @&#8203;@&#8203;
&nbsp;                        Perl\_warner\(aTHX\_ packWARN\(WARN\_SIGNAL\)\, "No
such signal&#8203;: SIG%s"\, s\);
&nbsp;                return 0;
&nbsp;            \}
\+\#ifdef HAS\_SIGPROCMASK
\+          /\* Avoid having the signal arrive at a bad time\, if possible\.
\*/
\+          sigemptyset\(&set\);
\+          sigaddset\(&set\,i\);
\+          sigprocmask\(SIG\_BLOCK\, &set\, &save\);
\+\#endif
 \#if defined\(FAKE\_PERSISTENT\_SIGNAL\_HANDLERS\) ||
defined\(FAKE\_DEFAULT\_SIGNAL\_HANDLERS\)
&nbsp;            if \(\!sig\_handlers\_initted\) Perl\_csighandler\_init\(\);
 \#endif
@&#8203;@&#8203; \-1189\,17 \+1232\,23 @&#8203;@&#8203;
&nbsp;              sig\_defaulting\[i\] = 0;
 \#endif
&nbsp;            SvREFCNT\_dec\(PL\_psig\_name\[i\]\);
\-           SvREFCNT\_dec\(PL\_psig\_ptr\[i\]\);
\+          to\_dec = PL\_psig\_ptr\[i\];
&nbsp;            PL\_psig\_ptr\[i\] = SvREFCNT\_inc\(sv\);
&nbsp;            SvTEMP\_off\(sv\); /\* Make sure it doesn't go away on us \*/
&nbsp;            PL\_psig\_name\[i\] = newSVpvn\(s\, len\);
&nbsp;            SvREADONLY\_on\(PL\_psig\_name\[i\]\);
&nbsp;     \}
&nbsp;     if \(SvTYPE\(sv\) == SVt\_PVGV || SvROK\(sv\)\) \{
\-           if \(i\)
\+          if \(i\) \{
&nbsp;                \(void\)rsignal\(i\, &Perl\_csighandler\);
\+\#ifdef HAS\_SIGPROCMASK
\+              sigprocmask\(SIG\_SETMASK\, &save\, NULL\);
\+\#endif
\+          \}
&nbsp;            else
&nbsp;                \*svp = SvREFCNT\_inc\(sv\);
\+          if\(to\_dec\)
\+              SvREFCNT\_dec\(to\_dec\);
&nbsp;            return 0;
&nbsp;     \}
&nbsp;     s = SvPV\_force\(sv\,len\);
@&#8203;@&#8203; \-1212\,7 \+1261\,7 @&#8203;@&#8203;
&nbsp;                \(void\)rsignal\(i\, SIG\_IGN\);
 \#endif
&nbsp;            \} else
\-               \*svp = 0;
\+              \*svp = 0; /\* XXX&#8203;: Redundant? \*/
&nbsp;     \}
&nbsp;     else if \(strEQ\(s\,"DEFAULT"\) || \!\*s\) \{
&nbsp;            if \(i\)
@&#8203;@&#8203; \-1225\,7 \+1274\,7 @&#8203;@&#8203;
&nbsp;                \(void\)rsignal\(i\, SIG\_DFL\);
 \#endif
&nbsp;            else
\-               \*svp = 0;
\+              \*svp = 0; /\* XXX&#8203;: Redundant? \*/
&nbsp;     \}
&nbsp;     else \{
&nbsp;            /\*
@&#8203;@&#8203; \-1240\,6 \+1289\,12 @&#8203;@&#8203;
&nbsp;            else
&nbsp;                \*svp = SvREFCNT\_inc\(sv\);
&nbsp;     \}
\+\#ifdef HAS\_SIGPROCMASK
\+    if\(i\)
\+          sigprocmask\(SIG\_SETMASK\, &save\, NULL\);
\+\#endif
\+    if\(to\_dec\)
\+          SvREFCNT\_dec\(to\_dec\);
&nbsp;     return 0;
 \}
 \#endif /\* \!PERL\_MICRO \*/
<details><summary>Inline Patch</summary>

```diff
diff -Naur perl-5.8.0/t/op/magic.t perl-5.8.0-sig2/t/op/magic.t
--- perl-5.8.0/t/op/magic.t            Wed Jul 10 17:18:07 2002
+++ perl-5.8.0-sig2/t/op/magic.t  Wed Jan 29 13:00:40 2003
@@ -36,7 +36,7 @@
     return 1;
 }
 
-print "1..46\n";
+print "1..48\n";
 
 $Is_MSWin32 = $^O eq 'MSWin32';
 $Is_NetWare = $^O eq 'NetWare';
@@ -67,7 +67,7 @@
 close FOO; # just mention it, squelch used-only-once
 
 if ($Is_MSWin32 || $Is_NetWare || $Is_Dos || $Is_MPE || $Is_MacOS) {
-    skip('SIGINT not safe on this platform') for 1..2;
+    skip('SIGINT not safe on this platform') for 1..4;
 }
 else {
   # the next tests are done in a subprocess because sh spits out a
@@ -98,7 +98,35 @@
 
     close CMDPIPE;
 
-    $test += 2;
+    open( CMDPIPE, "| $PERL");
+    print CMDPIPE <<'END';
+
+    { package X;
+          sub DESTROY {
+              kill "INT",$$;
+          }
+    }
+    sub x {
+          my $x=bless [], 'X';
+          return sub { $x };
+    }
+    $| = 1;                     # command buffering
+    $SIG{"INT"} = "ok5";
+    {
+          local $SIG{"INT"}=x();
+          print ""; # Needed to expose failure in 5.8.0 (why?)
+    }
+    sleep 1;
+    delete $SIG{"INT"};
+    kill "INT",$$; sleep 1;
+    sub ok5 {
+          print "ok 5\n";
+    }
+END
+    close CMDPIPE;
+    print $? & 0xFF ? "ok 6\n" : "not ok 6\n";
+
+    $test += 4;
 }
 
 # can we slice ENV?
 

```
</details>
Here is the output from "perlbug -d":
 


---
Flags&#8203;:
&nbsp;    category=
&nbsp;    severity=


---
Site configuration information for perl v5\.8\.0&#8203;:
 
Configured by anders at Fri Nov 15 12&#8203;:33&#8203;:55 PST 2002\.
 
Summary of my perl5 \(revision 5\.0 version 8 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.4\.18\, archname=i686\-linux
&nbsp;    uname='linux wolf12 2\.4\.18 \#1 mon jul 8 14&#8203;:56&#8203;:03 pdt 2002 i686
unknown '
&nbsp;    config\_args='\-de'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef
usemultiplicity=undef
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-strict\-aliasing \-I/usr/local/include
\-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64 \-I/usr/include/gdbm'\,
&nbsp;    optimize='\-O2'\,
&nbsp;    cppflags='\-fno\-strict\-aliasing \-I/usr/local/include
\-I/usr/include/gdbm'
&nbsp;    ccversion=''\, gccversion='2\.96 20000731 \(Red Hat Linux 7\.1
2\.96\-98\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\,
lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =' \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /lib /usr/lib
&nbsp;    libs=\-lnsl \-lndbm \-lgdbm \-ldl \-lm \-lc \-lcrypt \-lutil
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lc \-lcrypt \-lutil
&nbsp;    libc=/lib/libc\-2\.2\.4\.so\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version='2\.2\.4'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-rdynamic'
&nbsp;    cccdlflags='\-fpic'\, lddlflags='\-shared \-L/usr/local/lib'
 
Locally applied patches&#8203;:
&nbsp;    
 


---
@&#8203;INC for perl v5\.8\.0&#8203;:
&nbsp;    /tools/By\-Install/perl/5\.8\.0/i686\-linux2\.4/1/lib/5\.8\.0/i686\-linux
&nbsp;    /tools/By\-Install/perl/5\.8\.0/i686\-linux2\.4/1/lib/5\.8\.0
&nbsp;    /tools/By\-Install/cpan/living/i686\-linux2\.4/1/lib/cpan/i686\-linux
&nbsp;    /tools/By\-Install/cpan/living/i686\-linux2\.4/1/lib/cpan
&nbsp;    /tools/By\-Install/cpan/living/i686\-linux2\.4/1/lib/cpan
&nbsp;    \.
 


---
Environment for perl v5\.8\.0&#8203;:
&nbsp;    HOME=/home1/anders
&nbsp;    LANG=en\_US
&nbsp;    LANGUAGE \(unset\)
 
LD\_LIBRARY\_PATH=/opt/cadence/linux/ldv\-3\.4/tools/lib&#8203;:/opt/cadence/linux/
ldv\-3\.4/tools/inca/lib&#8203;:/usr/local/lib
&nbsp;    LOGDIR \(unset\)
 
PATH=/home1/anders/bin&#8203;:/tools/bin&#8203;:/opt/cadence/linux/ldv\-3\.4/tools/bin&#8203;:/
opt/cadence/linux/ldv\-3\.4/tools/inca/bin&#8203;:/opt/synopsys/linux/vcs/bin&#8203;:/op
t/synopsys/linux/vcs/virsimdir/bin&#8203;:/usr/local/vnc&#8203;:/opt/synopsys/linux/sc
l/linux/bin&#8203;:/usr/local/bin&#8203;:/bin&#8203;:/usr/bin&#8203;:/usr/X11R6/bin&#8203;:\.
&nbsp;    PERLDOC\_PAGER=less \-Cr
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/csh

**Comment by p5pRT** [comment]
### From goldbb2@earthlink.net

"Anders Johnson \(via RT\)" wrote&#8203;:
\[snip\]
> \+              if\(PL\_psig\_ptr\[i\]\) \{
> \+                      to\_dec=PL\_psig\_ptr\[i\];
> \+                      PL\_psig\_ptr\[i\]=0;
> \+                      SvREFCNT\_dec\(to\_dec\);
> \+              \}

How often is stuff like this used/needed\, throughout perl's source?
It's not exactly something one can grep for\, after all\.\.\.

I suspect \(fear\) that many places where this should be done\, but isn't\.

Eg\, av\_clear stores sv\_undef in the array slot before decrementing the
refcount of that sv\, but neither av\_undef nor av\_fill do\.

In gv\_fetchmeth\, I see&#8203;:
&nbsp;            SvREFCNT\_dec\(cv\);
&nbsp;            GvCV\(topgv\) = cv = Nullcv;
Should this do something similar \(replace\, then dec?\)?  I suppose that
it's rare that coderefs get blessed at all\, let alone into classes with
destructors\, so it would be hard to provoke this accidentally\, but it's
concievable\.  There're other places throughout perl which likewise
derecemnt before replacing\.

&nbsp;   perl \-e '@&#8203;x = \(bless\[\]\); sub DESTROY\{pop @&#8203;x\}; undef @&#8203;x'
&nbsp;   perl \-e '@&#8203;x = \(bless\[\]\); sub DESTROY\{pop @&#8203;x\}; $\#x=\-1'
&nbsp;   perl \-e '%x = \(1\,bless\[\]\); sub DESTROY\{ delete $x\{1\}; \}; %x=\(\)'

Knowing that the problem exists\, I'm sure that someone could come up
with perl code which provokes the others\.

\-\- 
$\.\.='\(?&#8203;:\(?\{local$^C=$^C|'\.\(1\<\<$\_\)\.'\}\)|\)'for\+a\.\.4;
$\.\.='\(?\{print\+substr"\\n \!\,$^C\,1 if $^C\<26\}\)\(?\!\)';
$\.=~s'\!'haktrsreltanPJ\,r  coeueh"';BEGIN\{$\{"\\cH"\}
|=\(1\<\<21\)\}""=~$\.;qw\(Just another Perl hacker\,\\n\);

**Comment by p5pRT** [comment]
### From ajohnson@wischip.com

Just so you know\, I have some revisions on this patch to deal with some
other potential problems that I just discovered by digging though mg\.c\.

Rather than confuse the issue right now\, I'll wait for confirmation that
I'm on the right track before submitting the revised patch\.

Thanks\,
&ers

**Comment by p5pRT** [comment]
### From ajohnson@wischip.com

I think I see your point\. Still\, my vote would be to fix these issues to
the extent that we know about them\. At worst\, there are 442 instances of
SvREFCNT\_dec to look at in the entire source tree\.

OTOH\, you can certainly make the argument that most of these bugs
require truly contrived cases to provoke\, and therefore it isn't worth
the effort\. However\, the particular bug fixed by the patch was provoked
by an attempt to accumulate received signals in a critical section
block\, and then deliver those signals to the process when the critical
section exits\, which I think you'll agree is legitimate \(but see below\)\.

Incidentally\, there are some related issues that I've recently
discovered&#8203;:

\+ If a signal is received just as you're changing its disposition from a
handler to DEFAULT\, then the signal could get deferred until after the
disposition changes\, which generally results in badness\. \(I already have
a patch for this one\.\)

\+ It's still not safe for a signal handler to die\(\) \(which is otherwise
very useful\)\, because it can prevent a DESTROY method from doing its
job\. \(This is analogous to the "destructors must never\, ever throw" rule
in C\+\+\. See http&#8203;://www\.gotw\.ca/gotw/047\.htm \.\)

I think that fixing the second issue requires adding some
critical\-section support to Perl itself\. \(Perhaps a $^E variable that
inhibits signal sampling and is automatically set locally for any
DESTROY and STORE methods that get called on behalf of a block exiting\.\)
If such a thing were to exist\, then you could certainly call my test
case contrived\, since you could then use "local $^E=1;" to accomplish
basically the same thing\.

Of course\, I'm all ears if you have any better ideas\. I'm particularly
fond of any solution that doesn't require modifying the Perl core\, so
long as it's clean and robust\.

Thanks\,
&ers

**Comment by p5pRT** [comment]
### From nick@ing-simmons.net

Benjamin Goldberg \<goldbb2@&#8203;earthlink\.net> writes&#8203;:
>"Anders Johnson \(via RT\)" wrote&#8203;:
>\[snip\]
>> \+              if\(PL\_psig\_ptr\[i\]\) \{
>> \+                      to\_dec=PL\_psig\_ptr\[i\];
>> \+                      PL\_psig\_ptr\[i\]=0;
>> \+                      SvREFCNT\_dec\(to\_dec\);
>> \+              \}
>
>How often is stuff like this used/needed\, throughout perl's source?
>It's not exactly something one can grep for\, after all\.\.\.
>
>I suspect \(fear\) that many places where this should be done\, but isn't\.

I don't fully grasp what the problem is yet\. The "safe signals" code 
in 5\.8 should make the above \_less\_ necessary\. Signals should only 
be asynchronously messing with a always allocated array of counts\.
All access to perl SVs etc\. should be synchronous \- so what is 
getting in during the decrement? And which bit of code let it?


>
>Eg\, av\_clear stores sv\_undef in the array slot before decrementing the
>refcount of that sv\, but neither av\_undef nor av\_fill do\.
>
>In gv\_fetchmeth\, I see&#8203;:
>            SvREFCNT\_dec\(cv\);
>            GvCV\(topgv\) = cv = Nullcv;
>Should this do something similar \(replace\, then dec?\)?  I suppose that
>it's rare that coderefs get blessed at all\, let alone into classes with
>destructors\, so it would be hard to provoke this accidentally\, but it's
>concievable\.  There're other places throughout perl which likewise
>derecemnt before replacing\.
>
>   perl \-e '@&#8203;x = \(bless\[\]\); sub DESTROY\{pop @&#8203;x\}; undef @&#8203;x'
>   perl \-e '@&#8203;x = \(bless\[\]\); sub DESTROY\{pop @&#8203;x\}; $\#x=\-1'
>   perl \-e '%x = \(1\,bless\[\]\); sub DESTROY\{ delete $x\{1\}; \}; %x=\(\)'
>
>Knowing that the problem exists\, I'm sure that someone could come up
>with perl code which provokes the others\.
\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
### From nick@ing-simmons.net

Anders Johnson \<ajohnson@&#8203;wischip\.com> writes&#8203;:
>Incidentally\, there are some related issues that I've recently
>discovered&#8203;:
>
>\+ If a signal is received just as you're changing its disposition from a
>handler to DEFAULT\, then the signal could get deferred until after the
>disposition changes\, which generally results in badness\. \(I already have
>a patch for this one\.\)

If the signal had arrived slightly later \(machine busier\)
it would have gone to DEFAULT handler \- so changing handlers when 
signals maybe in the pipe is source of the "badness"\.


>
>\+ It's still not safe for a signal handler to die\(\) \(which is otherwise
>very useful\)\, because it can prevent a DESTROY method from doing its
>job\. \(This is analogous to the "destructors must never\, ever throw" rule
>in C\+\+\. See http&#8203;://www\.gotw\.ca/gotw/047\.htm \.\)

Can you explain some more?


>
>I think that fixing the second issue requires adding some
>critical\-section support to Perl itself\. \(Perhaps a $^E variable that
>inhibits signal sampling and is automatically set locally for any
>DESTROY and STORE methods that get called on behalf of a block exiting\.\)
>If such a thing were to exist\, then you could certainly call my test
>case contrived\, since you could then use "local $^E=1;" to accomplish
>basically the same thing\.
>
>Of course\, I'm all ears if you have any better ideas\. I'm particularly
>fond of any solution that doesn't require modifying the Perl core\, so
>long as it's clean and robust\.
>
>Thanks\,
>&ers
\-\- 
Nick Ing\-Simmons
http&#8203;://www\.ni\-s\.u\-net\.com/

**Comment by p5pRT** [comment]
### From ajohnson@wischip.com

\[I tried to send this yesterday\, but our email server was hosed\. Sorry
if this is a duplicate\.\]

Nick Ing\-Simmons&#8203;:
> I don't fully grasp what the problem is yet\. The "safe signals" code 
> in 5\.8 should make the above \_less\_ necessary\. Signals should only 
> be asynchronously messing with a always allocated array of counts\. All
> access to perl SVs etc\. should be synchronous \- so what is 
> getting in during the decrement? And which bit of code let it?

This particular problems arises from the DESTROY method\(s\) associated
with the scalar being freed through SvREFCNT\_dec\. Signals get sampled
inside DESTROY methods\, so calls to SvREFCNT\_dec can result in
PERL\_ASYNC\_CHECK\(\) being called\. The test case included in the patch
illustrates the problem\.

The sigprocmask\(\) stuff in the original patch is actually bogus because
of the new "safe signals" code\. Sorry about that\. However\, something
similar is needed to fix one of the related problems \(see below\)\.

Nick Ing\-Simmons&#8203;:
> If the signal had arrived slightly later \(machine busier\)
> it would have gone to DEFAULT handler \- so changing handlers when 
> signals maybe in the pipe is source of the "badness"\.

Let me try to predict what will happen \(and since this is difficult to
reproduce reliably\, I'll let you correct me if I'm wrong\)&#8203;:

1\. We enter Perl\_magic\_setsig\(\) with a new disposition of "DEFAULT"\.
2\. The signal arrives\, setting PL\_sig\_pending and PL\_psig\_pend\[sig\]\.
3\. PL\_psig\_ptr\[sig\] and PL\_psig\_name\[sig\] are set to "DEFAULT"\.
4\. The signal is redispositioned to SIG\_DFL\.
5\. We return from Perl\_magic\_setsig\(\)\, and back into the Perl code\.
6\. At some point soon thereafter\, Perl\_despatch\_signal\(\) is called\,
which
&nbsp;   calls Perl\_sighandler\(\) on behalf of the signal\.
7\. "SIGxxx handler \\"DEFAULT\\" not defined\." \(I call this badness\.\)

It seems to me that fundamentally you need to prevent \#2 from happening\,
or you'll be in trouble\. You could require the Perl code to call
POSIX&#8203;::sigprocmask\(\) before assigning to %SIG\, but that's pretty ugly\.
My vote would be to have Perl\_magic\_setsig\(\) and Perl\_magic\_clearsig\(\)
take care of this for you\. \(I have a patch for this\, but I'd prefer to
resolve the issues before we start shooting various patches back and
forth\.\)

Nick Ing\-Simmons&#8203;:
> Can you explain some more?

That bit about destructors not throwing is a subtle point that C\+\+'ers
have debated at some length\, but ultimately the only clean solution is
that it should never\, ever happen\. Ever\. Here's an example&#8203;:

Consider an object that represents a temporary file\. When the object
goes out of scope\, you require the temporary file to be removed \*without
exception\*\. Now suppose that you just reached the end of the block
containing the object\, and you're in the middle of the object's DESTROY
method\, after you've done some stuff but before the file has been
removed\. Then a signal arrives and its handler die\(\)s\. Now you're in a
world of hurt\, because either A\) you've already decremented the object's
reference count\, and the file isn't going to be removed at all\, or B\)
the die\(\) causes the reference count to go to zero\, in which case you'll
restart the DESTROY method from the beginning\, which is bad because the
object is probably in an inconsistent state\.

There are lots of other ways for this to cause problems\, including
leaking gobs of memory and so forth\. But if you can take care of this
example\, then you're probably on the right track\.

I can think of two distinct ways to attack the problem\, the first of
which is that "$^E" stuff I mentioned earlier\. The second would be to
guarantee that DESTROY and STORE methods \(STORE can get called when a
tie'd local goes out of scope\) always get restarted\, and then to require
those methods to put a "local $SIG\{\.\.\.\}" or "eval" around critical code\.
\(This second approach leads us back to handlers that can accumulate
signals\, which is what exposes that first problem\.\)

Thanks\,
&ers

**Comment by p5pRT** [comment]
### From ajohnson@wischip.com

This thread seems to have died out unexpectedly\. I've included my latest
patch \(relative to 5\.8\.0\)\, which is alleged to do the following&#8203;:

\+ Signal handler isn't freed until after the new disposition is
installed\.
\+ Signal can't get deferred and then polled after the handler is
removed\.
\+ delete $SIG\{\.\.\.\} is roughly equivalent to $SIG\{\.\.\.\}="DEFAULT"\.
\+ A couple of redundant statements were removed\.

There is still the issue of a signal handler calling die\(\) while the
main thread is inside a destructor\. I'd like to discuss what can be done
about this\, but it shouldn't necessarily preclude applying the patch
as\-is\.

Robust\, clean signal handling is apparently a goal of Perl 5\.7\.3 and
later\. I think it's important to make that a reality\, and not just a
promise\. I'm here to help if I can\.

Thanks\,
&ers

\-\-\-\-\- Begin Patch \-\-\-\-\-
<details><summary>Inline Patch</summary>

```diff
diff -Naur perl-5.8.0/mg.c perl-5.8.0-sig3/mg.c
--- perl-5.8.0/mg.c	Sat Jun 15 13:16:44 2002
+++ perl-5.8.0-sig3/mg.c	Wed Jan 29 17:17:46 2003
@@ -1028,6 +1028,14 @@
 #endif
 
 #ifndef PERL_MICRO
+#ifdef HAS_SIGPROCMASK
+static void
+restore_sigmask(pTHX_ SV *save_sv)
+{
+    sigset_t *ossetp = (sigset_t *) SvPV_nolen( save_sv );
+    (void)sigprocmask(SIG_SETMASK, ossetp, (sigset_t *)0);
+}
+#endif
 int
 Perl_magic_getsig(pTHX_ SV *sv, MAGIC *mg)
 {
@@ -1061,19 +1069,67 @@
 int
 Perl_magic_clearsig(pTHX_ SV *sv, MAGIC *mg)
 {
-    I32 i;
+    /* XXX Some of this code was copied from Perl_magic_setsig. A

```
</details>
little
\+     \* refactoring might be in order\.
\+     \*/
\+    register char \*s;
&nbsp;     STRLEN n\_a;
\-    /\* Are we clearing a signal entry? \*/
\-    i = whichsig\(MgPV\(mg\,n\_a\)\);
\-    if \(i\) \{
\-        if\(PL\_psig\_ptr\[i\]\) \{
\-            SvREFCNT\_dec\(PL\_psig\_ptr\[i\]\);
\-            PL\_psig\_ptr\[i\]=0;
\-        \}
\-        if\(PL\_psig\_name\[i\]\) \{
\-            SvREFCNT\_dec\(PL\_psig\_name\[i\]\);
\-            PL\_psig\_name\[i\]=0;
\-        \}
\+    SV\* to\_dec;
\+    s = MgPV\(mg\,n\_a\);
\+    if \(\*s == '\_'\) \{
\+    SV\*\* svp;
\+    if \(strEQ\(s\,"\_\_DIE\_\_"\)\)
\+        svp = &PL\_diehook;
\+    else if \(strEQ\(s\,"\_\_WARN\_\_"\)\)
\+        svp = &PL\_warnhook;
\+    else
\+        Perl\_croak\(aTHX\_ "No such hook&#8203;: %s"\, s\);
\+    if \(\*svp\) \{
\+        to\_dec = \*svp;
\+        \*svp = 0;
\+            SvREFCNT\_dec\(to\_dec\);
\+    \}
\+    \}
\+    else \{
\+    I32 i;
\+    /\* Are we clearing a signal entry? \*/
\+    i = whichsig\(s\);
\+    if \(i\) \{
\+\#ifdef HAS\_SIGPROCMASK
\+        sigset\_t set\, save;
\+        SV\* save\_sv;
\+        /\* Avoid having the signal arrive at a bad time\, if
possible\. \*/
\+        sigemptyset\(&set\);
\+        sigaddset\(&set\,i\);
\+        sigprocmask\(SIG\_BLOCK\, &set\, &save\);
\+        ENTER;
\+        save\_sv = newSVpv\(\(char \*\)\(&save\)\, sizeof\(sigset\_t\)\);
\+        SAVEFREESV\(save\_sv\);
\+        SAVEDESTRUCTOR\_X\(restore\_sigmask\, save\_sv\);
\+\#endif
\+        PERL\_ASYNC\_CHECK\(\);
\+\#if defined\(FAKE\_PERSISTENT\_SIGNAL\_HANDLERS\) ||
defined\(FAKE\_DEFAULT\_SIGNAL\_HANDLERS\)
\+        if \(\!sig\_handlers\_initted\) Perl\_csighandler\_init\(\);
\+\#endif
\+\#ifdef FAKE\_DEFAULT\_SIGNAL\_HANDLERS
\+        sig\_defaulting\[i\] = 1;
\+        \(void\)rsignal\(i\, &Perl\_csighandler\);
\+\#else
\+        \(void\)rsignal\(i\, SIG\_DFL\);
\+\#endif
\+            if\(PL\_psig\_name\[i\]\) \{
\+            SvREFCNT\_dec\(PL\_psig\_name\[i\]\);
\+            PL\_psig\_name\[i\]=0;
\+            \}
\+            if\(PL\_psig\_ptr\[i\]\) \{
\+        to\_dec=PL\_psig\_ptr\[i\];
\+            PL\_psig\_ptr\[i\]=0;
\+        LEAVE;
\+            SvREFCNT\_dec\(to\_dec\);
\+            \}
\+        else
\+        LEAVE;
\+    \}
&nbsp;     \}
&nbsp;     return 0;
 \}
@&#8203;@&#8203; \-1156\,7 \+1212\,16 @&#8203;@&#8203;
&nbsp;     register char \*s;
&nbsp;     I32 i;
&nbsp;     SV\*\* svp = 0;
\+    /\* Need to be careful with SvREFCNT\_dec\(\)\, because that can have
side
\+     \* effects \(due to closures\)\. We must make sure that the new
disposition
\+     \* is in place before it is called\.
\+     \*/
\+    SV\* to\_dec = 0;
&nbsp;     STRLEN len;
\+\#ifdef HAS\_SIGPROCMASK
\+    sigset\_t set\, save;
\+    SV\* save\_sv;
\+\#endif
 
&nbsp;     s = MgPV\(mg\,len\);
&nbsp;     if \(\*s == '\_'\) \{
@&#8203;@&#8203; \-1168\,7 \+1233\,7 @&#8203;@&#8203;
&nbsp;         Perl\_croak\(aTHX\_ "No such hook&#8203;: %s"\, s\);
&nbsp;     i = 0;
&nbsp;     if \(\*svp\) \{
\-        SvREFCNT\_dec\(\*svp\);
\+        to\_dec = \*svp;
&nbsp;         \*svp = 0;
&nbsp;     \}
&nbsp;     \}
@&#8203;@&#8203; \-1179\,6 \+1244\,17 @&#8203;@&#8203;
&nbsp;         Perl\_warner\(aTHX\_ packWARN\(WARN\_SIGNAL\)\, "No such
signal&#8203;: SIG%s"\, s\);
&nbsp;         return 0;
&nbsp;     \}
\+\#ifdef HAS\_SIGPROCMASK
\+    /\* Avoid having the signal arrive at a bad time\, if possible\. \*/
\+    sigemptyset\(&set\);
\+    sigaddset\(&set\,i\);
\+    sigprocmask\(SIG\_BLOCK\, &set\, &save\);
\+    ENTER;
\+    save\_sv = newSVpv\(\(char \*\)\(&save\)\, sizeof\(sigset\_t\)\);
\+    SAVEFREESV\(save\_sv\);
\+    SAVEDESTRUCTOR\_X\(restore\_sigmask\, save\_sv\);
\+\#endif
\+    PERL\_ASYNC\_CHECK\(\);
 \#if defined\(FAKE\_PERSISTENT\_SIGNAL\_HANDLERS\) ||
defined\(FAKE\_DEFAULT\_SIGNAL\_HANDLERS\)
&nbsp;     if \(\!sig\_handlers\_initted\) Perl\_csighandler\_init\(\);
 \#endif
@&#8203;@&#8203; \-1186\,20 \+1262\,26 @&#8203;@&#8203;
&nbsp;     sig\_ignoring\[i\] = 0;
 \#endif
 \#ifdef FAKE\_DEFAULT\_SIGNAL\_HANDLERS
\-      sig\_defaulting\[i\] = 0;
\+    sig\_defaulting\[i\] = 0;
 \#endif
&nbsp;     SvREFCNT\_dec\(PL\_psig\_name\[i\]\);
\-    SvREFCNT\_dec\(PL\_psig\_ptr\[i\]\);
\+    to\_dec = PL\_psig\_ptr\[i\];
&nbsp;     PL\_psig\_ptr\[i\] = SvREFCNT\_inc\(sv\);
&nbsp;     SvTEMP\_off\(sv\); /\* Make sure it doesn't go away on us \*/
&nbsp;     PL\_psig\_name\[i\] = newSVpvn\(s\, len\);
&nbsp;     SvREADONLY\_on\(PL\_psig\_name\[i\]\);
&nbsp;     \}
&nbsp;     if \(SvTYPE\(sv\) == SVt\_PVGV || SvROK\(sv\)\) \{
\-    if \(i\)
\+    if \(i\) \{
&nbsp;         \(void\)rsignal\(i\, &Perl\_csighandler\);
\+\#ifdef HAS\_SIGPROCMASK
\+        LEAVE;
\+\#endif
\+    \}
&nbsp;     else
&nbsp;         \*svp = SvREFCNT\_inc\(sv\);
\+    if\(to\_dec\)
\+        SvREFCNT\_dec\(to\_dec\);
&nbsp;     return 0;
&nbsp;     \}
&nbsp;     s = SvPV\_force\(sv\,len\);
@&#8203;@&#8203; \-1211\,8 \+1293\,7 @&#8203;@&#8203;
 \#else
&nbsp;         \(void\)rsignal\(i\, SIG\_IGN\);
 \#endif
\-    \} else
\-        \*svp = 0;
\+    \}
&nbsp;     \}
&nbsp;     else if \(strEQ\(s\,"DEFAULT"\) || \!\*s\) \{
&nbsp;     if \(i\)
@&#8203;@&#8203; \-1224\,8 \+1305\,6 @&#8203;@&#8203;
 \#else
&nbsp;         \(void\)rsignal\(i\, SIG\_DFL\);
 \#endif
\-    else
\-        \*svp = 0;
&nbsp;     \}
&nbsp;     else \{
&nbsp;     /\*
@&#8203;@&#8203; \-1240\,6 \+1319\,12 @&#8203;@&#8203;
&nbsp;     else
&nbsp;         \*svp = SvREFCNT\_inc\(sv\);
&nbsp;     \}
\+\#ifdef HAS\_SIGPROCMASK
\+    if\(i\)
\+    LEAVE;
\+\#endif
\+    if\(to\_dec\)
\+    SvREFCNT\_dec\(to\_dec\);
&nbsp;     return 0;
 \}
 \#endif /\* \!PERL\_MICRO \*/
<details><summary>Inline Patch</summary>

```diff
diff -Naur perl-5.8.0/t/op/magic.t perl-5.8.0-sig3/t/op/magic.t
--- perl-5.8.0/t/op/magic.t	Wed Jul 10 17:18:07 2002
+++ perl-5.8.0-sig3/t/op/magic.t	Wed Jan 29 13:00:40 2003
@@ -36,7 +36,7 @@
     return 1;
 }
 
-print "1..46\n";
+print "1..48\n";
 
 $Is_MSWin32 = $^O eq 'MSWin32';
 $Is_NetWare = $^O eq 'NetWare';
@@ -67,7 +67,7 @@
 close FOO; # just mention it, squelch used-only-once
 
 if ($Is_MSWin32 || $Is_NetWare || $Is_Dos || $Is_MPE || $Is_MacOS) {
-    skip('SIGINT not safe on this platform') for 1..2;
+    skip('SIGINT not safe on this platform') for 1..4;
 }
 else {
   # the next tests are done in a subprocess because sh spits out a
@@ -98,7 +98,35 @@
 
     close CMDPIPE;
 
-    $test += 2;
+    open( CMDPIPE, "| $PERL");
+    print CMDPIPE <<'END';
+
+    { package X;
+	sub DESTROY {
+	    kill "INT",$$;
+	}
+    }
+    sub x {
+	my $x=bless [], 'X';
+	return sub { $x };
+    }
+    $| = 1;		# command buffering
+    $SIG{"INT"} = "ok5";
+    {
+	local $SIG{"INT"}=x();
+	print ""; # Needed to expose failure in 5.8.0 (why?)
+    }
+    sleep 1;
+    delete $SIG{"INT"};
+    kill "INT",$$; sleep 1;
+    sub ok5 {
+	print "ok 5\n";
+    }
+END
+    close CMDPIPE;
+    print $? & 0xFF ? "ok 6\n" : "not ok 6\n";
+
+    $test += 4;
 }
 
 # can we slice ENV?
```
</details>

**Comment by p5pRT** [comment]
### From @hvds

"Anders Johnson" \<ajohnson@&#8203;wischip\.com> wrote&#8203;:
:This thread seems to have died out unexpectedly\. I've included my latest
:patch \(relative to 5\.8\.0\)\, which is alleged to do the following&#8203;:
:
:\+ Signal handler isn't freed until after the new disposition is
:installed\.
:\+ Signal can't get deferred and then polled after the handler is
:removed\.
:\+ delete $SIG\{\.\.\.\} is roughly equivalent to $SIG\{\.\.\.\}="DEFAULT"\.
:\+ A couple of redundant statements were removed\.
:
:There is still the issue of a signal handler calling die\(\) while the
:main thread is inside a destructor\. I'd like to discuss what can be done
:about this\, but it shouldn't necessarily preclude applying the patch
:as\-is\.

Thanks\, applied to the development sources as change \#18803\.

Please note that your mailer appears to be wrapping lines in patches;
if you don't have a way to stop it doing that\, I'd suggest sending
any future patches as attachments instead\.

Hugo

**Comment by p5pRT** [comment]
@rspier - Status changed from 'new' to 'resolved'
